src/ray/metrics/perf_counter.h

```diff
@@ -0,0 +1,142 @@
+#ifndef _RAY_PERF_COUNTER_H_
+#define _RAY_PERF_COUNTER_H_
+
+#include <string>
+
+#include "opencensus/exporters/stats/prometheus/prometheus_exporter.h"
+#include "prometheus/exposer.h"
+#include "opencensus/tags/tag_key.h"
+#include "opencensus/stats/stats.h"
+
+namespace ray {
+
+namespace perf_counter {
+
+/// Include tag_def.h to define tag items
+#define RAY_TAG(name)                                                       \
+  static opencensus::tags::TagKey name##Key() {                             \
+    static const auto key = opencensus::tags::TagKey::Register(#name);      \
+    return key;                                                             \
+  }
+
+#include "tag_def.h"
+#undef RAY_TAG
+
+
+/// A thin wrapper that wraps the `opencensus::tag::measure` for using it simply.
+class Metric final {
+
+ public:
+  Metric(const std::string &name,
+         const std::string &description,
+         const std::string &unit) {
+    this->measure_ = std::make_shared<opencensus::stats::Measure<double>>(
+        opencensus::stats::Measure<double>::Register(
+            name, description, unit));
+  }
+
+  ~Metric() = default;
+
+  std::string GetName() const {
+    return measure_->GetDescriptor().name();
+  }
+
+  void Record(double value) {
+    Record(value, {});
+  }
+
+  void Record(double value, const std::vector<std::pair<opencensus::tags::TagKey::TagKey, std::string>>& tags) {
+    // global tags should be registered here.
+    static std::vector<std::pair<opencensus::tags::TagKey, std::string>> global_tags = {
+        {ray::perf_counter::JobNameKey(), "raylet"}
+    };
+
+    std::vector<std::pair<opencensus::tags::TagKey, std::string>> combined_tags(tags);
+    combined_tags.insert(std::end(combined_tags), std::begin(global_tags), std::end(global_tags));
+    opencensus::stats::Record({{*this->measure_, value}}, combined_tags);
+  }
+
+ private:
+  std::shared_ptr<opencensus::stats::Measure<double>> measure_ = nullptr;
+
+}; // class Metric
+
+/// Include metric_def.h to define tag items
+#define RAY_METRIC(name, description, unit)                                 \
```

Comment:
Yep, it depends on which kind of usage we need, `RedisLatency.Record()` or `RedisLatency().Record().`

As @raulchen suggestion in the [doc](https://docs.google.com/document/d/1bC6JTsw67DLiKbMhf2jaIG8_-gW5O5G1-vD4fS7fDuI/edit#),  I assume that he prefer to `RedisLatency().Record()`(function call). 

If we can all agree on `RedisLatency.Record()`, it does make sense to remove the macro and define metrics directly.

Reply:
Note that you can also can also define `Metric operator () ()` to achieve the desired parens if needed.

I'd prefer to avoid using macros when possible, according to https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros